File: App.client/projects/proctor-app/scripts/models/complaint.ts
import { User, ComplaintStatus, Priority } from './user';
import { ComplaintCategory } from './complaint-category';
import { ComplaintEvidence } from './complaint-evidence';
import { CaseAssignment } from './case-assignment';
import { Explanation } from './explanation';
import { CaseFile } from './case-file';
import { Meeting } from './meeting';

export class Complaint {
  id: number;
  title: string;
  description: string;
  complaintDate: string;
  status: string;
  priority: string;
  complainantId: number;
  complainantName?: string | null;
  complainantDetails?: string | null;
  complainantStudentId?: string | null;
  accusedStudentId?: number | null;
  accusedName?: string | null;
  accusedDetails?: string | null;
  categoryId?: number | null;
  location: string;
  incidentDate?: string | null;
  createdOn: string;
  modifiedOn: string;
  complainant?: User | null;
  category?: ComplaintCategory | null;
  evidence: ComplaintEvidence[];
  caseAssignments: CaseAssignment[];
  explanations: Explanation[];
  caseFiles: CaseFile[];
  meetings: Meeting[];

  constructor() {
    this.id = 0;
    this.title = this.description = this.location = '';
    this.complainantId = 0;
    this.complainantName = this.complainantDetails = this.complainantStudentId = null;
    this.accusedName = this.accusedDetails = null;
    this.status = ComplaintStatus.Pending;
    this.priority = Priority.Medium;
    this.complaintDate = new Date().toISOString();
    this.evidence = [];
    this.caseAssignments = [];
    this.explanations = [];
    this.caseFiles = [];
    this.meetings = [];
    this.createdOn = this.modifiedOn = '';
  }
}

File: App.Models/Complaint.cs
using App.Core.Models;

namespace App.Models
{
    public class Complaint : IAuditableEntity
    {
        public int Id { get; set; }
        public string Title { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public DateTime ComplaintDate { get; set; }
        public ComplaintStatus Status { get; set; } = ComplaintStatus.Pending;
        public Priority Priority { get; set; } = Priority.Medium;
        
        public int ComplainantId { get; set; }
        public string? ComplainantName { get; set; }
        public string? ComplainantDetails { get; set; }
        public string? ComplainantStudentId { get; set; }
        
        public int? AccusedStudentId { get; set; }
        public string? AccusedName { get; set; }
        public string? AccusedDetails { get; set; }
        public int? CategoryId { get; set; }
        public string? Location { get; set; }
        public DateTime? IncidentDate { get; set; }
        public DateTime CreatedOn { get; set; }
        public DateTime ModifiedOn { get; set; }

        // Navigation properties
        public virtual User Complainant { get; set; } = null!;
        public virtual ComplaintCategory? Category { get; set; }
        public virtual ICollection<ComplaintEvidence> Evidence { get; set; } = new List<ComplaintEvidence>();
        public virtual ICollection<CaseAssignment> CaseAssignments { get; set; } = new List<CaseAssignment>();
        public virtual ICollection<Explanation> Explanations { get; set; } = new List<Explanation>();
        public virtual ICollection<CaseFile> CaseFiles { get; set; } = new List<CaseFile>();
        public virtual ICollection<Meeting> Meetings { get; set; } = new List<Meeting>();
    }
}


